src/twisted/protocols/_smb/core.py

```diff
@@ -0,0 +1,519 @@
+        log.debug("Size             {sz}", sz=req.size)
+        log.debug("Security mode    0x{sm:08x}", sm=req.security_mode)
+        log.debug("Capabilities     0x{cap:08x}", cap=req.capabilities)
+        log.debug("Channel          0x{chl:08x}", chl=req.channel)
+        log.debug("Prev. session ID 0x{pid:016x}", pid=req.prev_session_id)
+        if self.first_session_setup:
+            self.blob_manager.receiveInitialBlob(blob)
+            resp.buffer = self.blob_manager.generateChallengeBlob()
+            self.session_setup_response(resp, STATUS_MORE_PROCESSING)
+            self.first_session_setup = False
+        else:
+            self.blob_manager.receiveResp(blob)
+            if self.blob_manager.credential:
+                log.debug("got credential: %r" % self.blob_manager.credential)
+                d = self.factory.portal.login(
+                    self.blob_manager.credential,
+                    SMBMind(req.prev_session_id,
+                            self.blob_manager.credential.domain, self.addr),
+                    ISMBServer)
+                d.addCallback(self._cb_login, resp)
+                d.addErrback(self._eb_login, resp)
+            else:
+                resp.buffer = self.blob_manager.generateChallengeBlob()
+                self.session_setup_response(resp, STATUS_MORE_PROCESSING)
+
+    def _cb_login(self, t, resp):
+        _, self.avatar, self.logout_thunk = t
+        resp.buffer = self.blob_manager.generateAuthResponseBlob(True)
+        log.debug("successful login")
+        self.session_setup_response(resp, STATUS_SUCCESS)
+
+    def _eb_login(self, failure, resp):
+        log.debug(failure.getTraceback())
+        resp.buffer = self.blob_manager.generateAuthResponseBlob(False)
+        self.session_setup_response(resp, STATUS_LOGON_FAILURE)
+
+    def session_setup_response(self, resp, ntstatus):
+        log.debug("session_setup_response")
+        resp.flags = 0
+        if self.blob_manager.credential == ANONYMOUS:
+            resp.flags |= SESSION_FLAG_IS_NULL
+        resp.size = 9
+        resp.offset = 72
+        resp.buflen = len(resp.buffer)
+        self.send_with_header(resp, 'session_setup', ntstatus)
+
+    def smb_logoff(self, req, resp):
+        assert req.size == 4
+        resp.size = 4
+        self.send_with_header(resp)
+        if self.logout_thunk:
+            d = maybeDeferred(self.logout_thunk)
+            d.addErrback(self._eb_logoff)
+
+    def _eb_logoff(self, f):
+        log.error(f.getTraceback())
+
+    def smb_tree_connect(self, req, resp):
+        if self.avatar is None:
+            self.error_response(STATUS_ACCESS_DENIED)
+            return
+        assert req.size == 9
+        path = req.buffer[req.offset - len(req) - 64:req.offset - len(req) -
+                          64 + req.buflen]
+        path = path.decode("utf-16le").split("\\")[-1]
+        try:
+            share = self.avatar.getShare(path)
+        except KeyError:
```

Comment:
Mmm… `KeyError` isn't a great error here, you probably want a custom exception. Also should this really be raised synchronously? This seems like the sort of thing that might want an async result...